### PR TITLE
#167750055 Build out Reset password Endpoints for Users

### PR DESCRIPTION
**Pull request by uthdev**
#### Description of Task to be completed?
 - write tests for the endpoints
 - Create an endpoint for sending a password reset email
 - create get endpoint for the password reset link
 - Create an endpoint for resetting the password by allowing a user to set a new password
 - Create client-side pre-commit and pre-push hooks
 #### How should this be manually tested?
- clone the repo
- cd into the root folder of the cloned repo
- run `npm install`  to install necessary dependencies
- run `npm run migrate` to create migrations
- run `npm run seed` to seed database with dummy users for test
- run `npm test` to run the test suite
- run `npm run start:dev` to start the server
- Go to Postman and make a POST request to `http://localhost:3000/api/v1/forgotpassword` sending `email: janedoe@example.com` in the request body
- You should get a link to reset user password if the user exists or an error otherwise
- Copy the link and make a POST request to the link sending the new password in the request body
- an appropriate response is sent to the user upon validation
- Run `./scripts/install-hooks.bash` to install the pre-commit and pre-push hooks
 #### Any background context you want to provide?
- Sample forgot password request
```
{
  "email": "fakemail@yahoo.com"
}
```
- Sample reset password request
```
{
  "password": "newPassword",
  "confirmPassword": "newPassword"
}
```
- The email sent in the request body is the seeded dummy user's email
- The link can only be used once to change password. Hence it becomes invalid after successful password reset
- The link sent expires in  a 24 hours period
 #### What are the relevant pivotal tracker stories?
[#167750055](https://www.pivotaltracker.com/story/show/167750055)
 #### Screenshots (if appropriate)
- Post endpoint for sending password reset email
Screenshot (134)
- Post endpoint for setting a new password
Screenshot (136)
- An attempt to use the same link to change the password
Screenshot (137)
 #### Questions:
Based on my implementation, the user gets to see the password reset form upon validation from the backend. I wonder if this should have been left to be handled from the frontend?

### CODE REVIEW

**Comment by darasimiolaifa** [comment]
Hi @uthdev Nice work here. I especially like your PR detail, plus reading your code taught me some nice tricks. Well done. But there are a few things I'd like you to look into.
The first is that the user model and migration you used for your code does not reflect the latest changes we made. I couldn't run your code because I would not like to create a migration that I would have to undo again.
Secondly is, there are a lot of eslint complaints that I think you could either fulfill their requirements or find a way around, or totally silence.
I'll like you to look into this. Once again, well done.

**Review by feobaby — COMMENTED** [review]
Hi @uthdev I kept trying and trying but the reset password link is forbidden, that is the messages I got.
Also you only have password as the an input field where resetting.
i think we also need an input field called confirmPassword

**so if (password !== confirmPassword){
// then soo, 
}**
This is typical in case of password match

**Comment by uthdev** [comment]
> Hi @uthdev I kept trying and trying but the reset password link is forbidden, that is the messages I got.
> Also you only have password as the an input field where resetting.
> i think we also need an input field called confirmPassword
> 
> **so if (password !== confirmPassword){ // then soo, }**
> This is typical in case of password match

Were you sent a link when you made a post request to the `/api/forgotpassword` endpoint?
And the validation issue will be handled once @Oliver-ke PR is merged. I had done what you are suggesting before in the controller just for the sake of the demo but removed it cause we should have all our validations in one place and not validating inputs inside the controllers.

**Comment by feobaby** [comment]
> > Hi @uthdev I kept trying and trying but the reset password link is forbidden, that is the messages I got.
> > Also you only have password as the an input field where resetting.
> > i think we also need an input field called confirmPassword
> > **so if (password !== confirmPassword){ // then soo, }**
> > This is typical in case of password match
> 
> Were you sent a link when you made a post request to the `/api/forgotpassword` endpoint?
> And the validation issue will be handled once @Oliver-ke PR is merged. I had done what you are suggesting before in the controller just for the sake of the demo but removed it cause we should have all our validations in one place and not validating inputs inside the controllers.

Okay now I understand. 
Yes, I was sent a link so when I copied that link to postman and tried posting a new password, it tells me password is invalid.
Although I send my password in JSON format, does it have anything to do with it. I really want to approve this request.

**Comment by uthdev** [comment]
> > > Hi @uthdev I kept trying and trying but the reset password link is forbidden, that is the messages I got.
> > > Also you only have password as the an input field where resetting.
> > > i think we also need an input field called confirmPassword
> > > **so if (password !== confirmPassword){ // then soo, }**
> > > This is typical in case of password match
> > 
> > 
> > Were you sent a link when you made a post request to the `/api/forgotpassword` endpoint?
> > And the validation issue will be handled once @Oliver-ke PR is merged. I had done what you are suggesting before in the controller just for the sake of the demo but removed it cause we should have all our validations in one place and not validating inputs inside the controllers.
> 
> Okay now I understand.
> Yes, I was sent a link so when I copied that link to postman and tried posting a new password, it tells me password is invalid.
> Although I send my password in JSON format, does it have anything to do with it. I really want to approve this request.

I've fixed it @funmi5. You can review now

**Review by feobaby — DISMISSED** [review]
Great job, keep it up.

**Comment by sa-ma** [comment]
You need to update your pr description. `npm migrate` and `npm seed` won't work

**Comment by uthdev** [comment]
> You need to update your pr description. `npm migrate` and `npm seed` won't work

Done @sa-ma

**Comment by sa-ma** [comment]
Fix eslint and test-coverage issues

File: src/controllers/resetPassword.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      const user = await models.users.findOne({
+        where: {
+          email
+        }
+      });
+      if (!user) {
+        return errorResponse(res, 404, 'No user with email');
+      }
+      const { id: userId, password: passwordHash, createdAt } = user;
+      const secret = `${passwordHash}-${createdAt}`;
```
Comment (eokwukwe) [inline src/controllers/resetPassword.js:33]: Use the secret from the `.env` file

File: src/controllers/resetPassword.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  static async setNewPassword(req, res) {
+    const { password } = req.body;
+    const { userId, token } = req.params;
+    try {
+      const { dataValues: user } = await models.users.findOne({
+        where: {
+          id: userId
+        }
+      });
+      const secret = `${user.password}-${user.createdAt}`;
```
Comment (eokwukwe) [inline src/controllers/resetPassword.js:60]: Use the secret from the `.env` file

File: src/controllers/resetPassword.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      const payload = await Jwt.verifyToken(token, secret);
+
+      if (payload.userId === user.id) {
+        const hashPassword = generateHash(password);
```
Comment (eokwukwe) [inline src/controllers/resetPassword.js]: Stick with the `hashPassword` function created by @oyewoas

File: src/routes/api/resetPassword.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+const {
+  sendPasswordResetEmail,
+  setNewPassword
+} = ResetPasswordController;
+
+const resetPasswordRouter = new Router();
+resetPasswordRouter.post('/forgotpassword', sendPasswordResetEmail);
+resetPasswordRouter.post('/resetpassword/:userId/:token', setNewPassword);
+
```
Comment (eokwukwe) [inline src/routes/api/resetPassword.js]: Make the token a query param: 
`/resetpassword/:userId/?token={token}`

File: src/routes/api/resetPassword.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+import ResetPasswordController from '../../controllers/resetPassword';
+
+const {
+  sendPasswordResetEmail,
+  setNewPassword
+} = ResetPasswordController;
+
+const resetPasswordRouter = new Router();
+resetPasswordRouter.post('/forgotpassword', sendPasswordResetEmail);
+resetPasswordRouter.post('/resetpassword/:userId/:token', setNewPassword);
```
Comment (eokwukwe) [inline src/routes/api/resetPassword.js]: Make the token a query params:
`/resetpassword/:userId/?token=resetToken`

**Review by eokwukwe — COMMENTED** [review]
@uthdev You need to consolidate the helper functions you created with the utils that are already existing.

File: src/controllers/resetPassword.js (hunk 1 above)
Comment (uthdev) [reply src/controllers/resetPassword.js:33]: On it

> @uthdev You need to consolidate the helper functions you created with the utils that are already existing.

Already doing that now @eokwukwe

**Review by oyewoas — COMMENTED** [review]
@uthdev well done, but please can you rebase this branch, and please i keep getting this error after sending a post request to the generated link, do you know what may cause this??
```json
{
    "status": 403,
    "error": "Password reset Link invalid"
}
```

**Comment by uthdev** [comment]
> @uthdev well done, but please can you rebase this branch, and please i keep getting this error after sending a post request to the generated link, do you know what may cause this??
> 
> ```json
> {
>     "status": 403,
>     "error": "Password reset Link invalid"
> }
> ```

@oyewoas Still working on resolving and consolidating the features. Most of the utils methods used in user registration are what I've implemented before your branch got merge so there is a lot of reshuffling to do. But I'll push soon

**Review by oyewoas — COMMENTED** [review]
welldone @uthdev, please update your PR Description, i needed to use 
```json
{
	"password":"passwords",
	"confirmPassword": "passwords"
}
``` 
In the request body for it to work, so that means both password and confirmPassword are needed in the `request body` for a successful password reset, this is because of the validation middleware that you added which verifies if the new password matched the confirmPassword.

**Review by oyewoas — DISMISSED** [review]

File: package.json [hunk 6]
```diff
@@ -4,16 +4,17 @@
   "description": "A travel booking server",
   "main": "index.js",
   "scripts": {
-    "pre-test": "cross-env NODE_ENV=test npm run migrate",
-    "test": "cross-env NODE_ENV=test nyc --reporter=html --reporter=text mocha src/test/index.test --require @babel/register --recursive --exit || true",
+    "pre-test": "cross-env NODE_ENV=test npm run undo-migrate && npm run migrate",
```
Comment (eokwukwe) [inline package.json]: why did you add `npm run undo-migrate`

File: src/index.js [hunk 7]
```diff
@@ -28,7 +28,7 @@ app.use((req, res, next) => {
 if (!isProduction) {
   app.use((err, req, res) => {
     // eslint-disable-next-line no-console
-    console.log(err.stack);
+    generateToken(err.stack);
```
Comment (eokwukwe) [inline src/index.js]: Why are you using `generateToken` here?

File: package.json (hunk 6 above)
Comment (uthdev) [reply package.json]: It was because of the changes made to the existing migrations. We had to undo all migrations and run migrate again.

File: src/index.js (hunk 7 above)
Comment (uthdev) [reply src/index.js]: That was a mistake. Did search and replace. I've fixed it

File: .gitignore [hunk 8]
```diff
@@ -48,6 +48,7 @@ build/Release
 # Dependency directories
 node_modules/
 jspm_packages/
+package-lock.json
```
Comment (sa-ma) [inline .gitignore]: I think this should be removed

**Comment by sa-ma** [comment]
The seeds are not running, validation error because firstName and lastName is required

**Comment by uthdev** [comment]
> The seeds are not running, validation error because firstName and lastName is required

They are set to allow null. And they running fine here @sa-ma

**Review by Oliver-ke — DISMISSED** [review]
Nice one @uthdev, works fine, I personally liked the way you checked for a used reset link

**Comment by uthdev** [comment]
> Nice one @uthdev, works fine, I personally liked the way you checked for a used reset link

Thanks @Oliver-ke

**Review by sa-ma — DISMISSED** [review]

File: src/routes/index.js [hunk 9]
```diff
@@ -1,7 +1,14 @@
...
 
-const router = express.Router();
+
+const router = new Router();
+router.get('/api/v1', (req, res) => {
+  res.status(200).json({
+    status: res.statusCode,
+    message: 'Welcome to Wolfsbane'
+  });
+});
```
Comment (eokwukwe) [inline src/routes/index.js]: Remove this block of code. We already have a similar routing for the base route in the `index.js `sever entry file

Comment (eokwukwe) [reply src/routes/index.js]: @uthdev remove the above block of code👆🏿

**Comment by eokwukwe** [comment]
@uthdev what you mean by: Run ./scripts/install-hooks.bash to install the pre-commit and pre-push hooks.
Are we running the script with npm

**Comment by uthdev** [comment]
> @uthdev what you mean by: Run ./scripts/install-hooks.bash to install the pre-commit and pre-push hooks

The hooks are not tracked by git as they are in the `.git` folder. So pushing to remote repository has to be done by anchoring it to a bash scripts @eokwukwe

**Comment by eokwukwe** [comment]
> > @uthdev what you mean by: Run ./scripts/install-hooks.bash to install the pre-commit and pre-push hooks
> 
> The hooks are not tracked by fit as they are in the `.git` folder. So pushing to remote repository has to be done by anchoring it to a bash scripts @eokwukwe

I don't understand. How do I run the script? What command do I need to type to run the script command?

**Comment by uthdev** [comment]
> > > @uthdev what you mean by: Run ./scripts/install-hooks.bash to install the pre-commit and pre-push hooks
> > 
> > 
> > The hooks are not tracked by fit as they are in the `.git` folder. So pushing to remote repository has to be done by anchoring it to a bash scripts @eokwukwe
> 
> I don't understand

We are not running with `npm`. We just run `./scripts/install-hooks.bash` in the terminal. The `install-hooks.bash` file inside the scripts folder has commands that links the pre-commit and pre-push scripts to the pre-commit and pre-push hooks that are inside the `.git/hooks` folder which are not tracked by git and cannot be pushed hence why they have to be anchored to scripts

**Comment by eokwukwe** [comment]
> ./scripits/install-hooks.bash

I am getting this error `no such file or directory: /scripits/install-hooks.bash`

**Comment by uthdev** [comment]
> > > @uthdev what you mean by: Run ./scripts/install-hooks.bash to install the pre-commit and pre-push hooks
> > 
> > 
> > The hooks are not tracked by fit as they are in the `.git` folder. So pushing to remote repository has to be done by anchoring it to a bash scripts @eokwukwe
> 
> I don't understand. How do I run the script? What command do I need to type to run the script command?

**Comment by uthdev** [comment]
> > > > @uthdev what you mean by: Run ./scripts/install-hooks.bash to install the pre-commit and pre-push hooks
> > > 
> > > 
> > > The hooks are not tracked by fit as they are in the `.git` folder. So pushing to remote repository has to be done by anchoring it to a bash scripts @eokwukwe
> > 
> > 
> > I don't understand. How do I run the script? What command do I need to type to run the script command?

`./scripts/install-hooks.bash`

**Review by oyewoas — APPROVED** [review]
welldone @uthdev , it works fine..
